substrate/frame/staking/rc-client/src/lib.rs

```diff
@@ -15,95 +15,268 @@
-
 use alloc::vec::Vec;
 use frame_support::pallet_prelude::*;
-use sp_core::crypto::AccountId32;
 use sp_runtime::Perbill;
 use sp_staking::SessionIndex;
-use xcm::prelude::*;
+
+/// Export everything needed for the pallet to be used in the runtime.
+pub use pallet::*;
 
 const LOG_TARGET: &str = "runtime::staking::rc-client";
 
-// Provides to the pallet a validator set produced by an election or other similar mechanism.
-pub trait ElectionResultHandler<ValidatorId> {
-	fn handle_election_result(result: Vec<ValidatorId>);
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: crate::LOG_TARGET,
+			concat!("[{:?}] ⬇️ ", $patter), <frame_system::Pallet<T>>::block_number() $(, $values)*
+		)
+	};
 }
 
-// API provided by the staking pallet.
-pub trait StakingApi {
-	/// New session with index `start_index` has started on the relay chain.
-	fn on_relay_chain_session_start(start_index: SessionIndex);
-	/// A session with index `end_index` has ended on the relay chain. The block authors and their
-	/// corresponding session points are reported.
-	fn on_relay_chain_session_end(end_index: SessionIndex, block_authors: Vec<(AccountId32, u32)>);
-	/// Report one or more offences on the relay chain.
-	fn on_new_offences(slash_session: SessionIndex, offences: Vec<Offence>) -> Weight;
-	/// Report validator rewards.
-	fn reward_by_ids(validators_points: Vec<(AccountId32, u32)>);
+/// The communication trait of `pallet-staking-rc-client` -> `relay-chain`.
+///
+/// This trait should only encapsulate our _outgoing_ communication to the RC. Any incoming
+/// communication comes it directly via our calls.
+///
+/// In a real runtime, this is implemented via XCM calls, much like how the coretime pallet works.
+/// In a test runtime, it can be wired to direct function call.
+pub trait SendToRelayChain {
+	/// The validator account ids.
+	type AccountId;
+
+	/// Send a new validator set report to relay chain.
+	fn validator_set(report: ValidatorSetReport<Self::AccountId>);
 }
 
-// TODO: update the comment
-/// Means for interacting with a specialized version of the `session` trait.
-///
-/// This is needed because `Staking` sets the `ValidatorIdOf` of the `pallet_session::Config`
-pub trait SessionInterface<AccountId> {
-	/// Get the validators from session.
-	fn validators() -> Vec<AccountId>;
+#[derive(Encode, Decode, DecodeWithMemTracking, Debug, Clone, PartialEq, TypeInfo)]
+/// A report about a new validator set. This is sent from AH -> RC.
+pub struct ValidatorSetReport<AccountId> {
+	/// The new validator set.
+	pub new_validator_set: Vec<AccountId>,
+	/// The id of this validator set.
+	///
+	/// Is an always incrementing identifier for this validator set, the activation of which can be
+	/// later pointed to in a `SessionReport`.
+	pub id: u32,
+	/// Signal the relay chain that it can prune up to this session, ans enough eras have passed.
```

Comment:
```suggestion
	/// Signal the relay chain that it can prune up to this session, ans enough eras have passed.
	///
	/// This can always have a safety buffer. For example, whatever is a sane value, it can be `value - 5`.
```